Engine/TestStepRun.cs

```diff
@@ -330,20 +330,31 @@ internal void ThrowDueToBreakConditions()
             throw new TestStepBreakException(TestStepName, Verdict);
         }
 
-        internal void WaitForOutput(OutputAvailability mode)
+        internal bool IsStepChildOf(ITestStep step, ITestStep possibleParent)
         {
+            do
+            {
+                step = step.Parent as ITestStep;
+            }
+            while (step != null && step != possibleParent);
+            return step != null;
+        }
+
+        internal void WaitForOutput(OutputAvailability mode, ITestStep waitingFor)
+        {
+            ITestStep currentStep = TestStepExtensions.currentlyExecutingTestStep;
             var currentThread = TapThread.Current;
             switch (mode)
             {
                 case OutputAvailability.BeforeRun: 
                     return;
                 case OutputAvailability.AfterDefer:
-                    if (StepThread == currentThread && runDone.Wait(0) == false)
+                    if ((StepThread == currentThread && runDone.Wait(0) == false) || IsStepChildOf(currentStep, waitingFor))
```

Comment:
We need to check if currentStep is null.

Reply:
I'm not sure how it is possible for currentStep to be null here, `WaitForOutput` only gets called from `DoRun`, where it is always set to a non null value before it is being run.
Even if we add a null check for the future im not even sure what should be done, how can a step wait for anything if the currently executing step is null?

Reply:
I think you have a good point. I can some up with two cases.

What if the test step starts a new thread, which waits for the input? Couldn't the test plan complete before the thread does it thing? 

Also user interfaces may read these values asynchronously, which could occur during the ru, but the thread static value is always null seen from the UI thread.

Reply:
Added a null check as requested